### PR TITLE
IMPRO-1354 use PySteps extrapolate method in CLI

### PR DESCRIPTION
**Pull request by MoseleyS**
Switches the extrapolation method in the nowcast-extrapolate CLI to PySteps.

Testing:
 - [x] Ran tests and they passed OK

### CODE REVIEW

**Comment by MoseleyS** [comment]
PySteps now uses the attributes_dict. I need to add the additional KGO file tomorrow. I've also added something to suppress the PySteps import message.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/stdout_trap.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/improver/improver/pullRequest?prid=4597667&bid=15433284)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/stdout_trap.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667&bid=15433284)

File: lib/improver/nowcasting/pysteps_advection.py [hunk 1]
```diff
@@ -72,8 +85,13 @@ def _get_precip_rate(self):
...
         """
-        self.analysis_cube, = ApplyOrographicEnhancement("subtract").process(
-            self.analysis_cube, self.orogenh)
+        if self.orogenh:
+            self.analysis_cube, = ApplyOrographicEnhancement(
+                "subtract").process(self.analysis_cube, self.orogenh)
+        elif "precipitation_rate" in self.analysis_cube.name():
+            msg = ("For precipitation fields, orographic enhancement "
+                   "cube must be supplied.")
+            raise ValueError(msg)
```
Comment (cgsandford) [inline lib/improver/nowcasting/pysteps_advection.py]: I'm going to say we don't need this here, at least not without some other changes.  The PySteps extrapolation code is designed to work with precipitation rates in mm/h, and this wrapper plugin has been written with that assumption embedded - this function for example is `_get_precip_rate`, and gets them in mm/h for pysteps' consumption.  I can't think of any situation in which we would not need an orographic enhancement cube input here.

File: lib/improver/utilities/stdout_trap.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Contains a class to trap stdout"""
```
Comment (cgsandford) [inline lib/improver/utilities/stdout_trap.py]: This works but I'd like Green team member to confirm they like the approach - @arh89 @benfitzpatrick ?

File: tests/improver-nowcast-extrapolate/04-basic_no_orographic_enhancement.bats [hunk 3]
```diff
@@ -33,7 +33,7 @@
 
 @test "extrapolate basic no orographic enhancement" {
   improver_check_skip_acceptance
-  KGO="nowcast-extrapolate/extrapolate_no_orographic_enhancement/kgo.nc"
+  KGO="nowcast-feature-branch/nowcast-extrapolate/kgo_no_orographic_enhancment.nc"
```
Comment (cgsandford) [inline tests/improver-nowcast-extrapolate/04-basic_no_orographic_enhancement.bats]: I don't think this test is relevant for the pysteps method.

File: lib/improver/nowcasting/pysteps_advection.py [hunk 4]
```diff
@@ -50,7 +55,7 @@ class PystepsExtrapolate(object):
         https://pysteps.readthedocs.io/en/latest/generated/
         pysteps.extrapolation.semilagrangian.extrapolate.html
     """
-    def __init__(self, interval, max_lead_time):
+    def __init__(self, interval, max_lead_time, attributes_dict=None):
```
Comment (cgsandford) [inline lib/improver/nowcasting/pysteps_advection.py]: Attributes dict I would usually pass into `process`.  This makes more sense in this particular instance as well, as `process` is where we're passing in anything specific to the outputs (as opposed to what the plugin does).  Then there's no need to make it a class member.

File: lib/improver/utilities/stdout_trap.py (hunk 2 above)
Comment (arh89) [reply lib/improver/utilities/stdout_trap.py]: Ugh, it's annoying that `pysteps` dumps random stuff to `stdout`!

I'm happy with the general approach of using a context manager - I'd suggest using `contextlib.contextmanager` to save creating a class in this way though. (If not, I probably wouldn't subclass from `list` though as I don't think that's appropriate.)

Also, for more generic code, you might want to pass what you want to capture as the input when you create the context manager. e.g: I could see use cases for capturing stdout and/or stderr as well.

Maybe @benfitzpatrick has other views though.

File: lib/improver/nowcasting/pysteps_advection.py [hunk 5]
```diff
@@ -34,13 +34,18 @@
...
 
+from improver.metadata.amend import (amend_attributes, set_history_attribute)
 from improver.utilities.spatial import (
     check_if_grid_is_equal_area, calculate_grid_spacing)
 from improver.utilities.temporal import (
     iris_time_to_datetime, datetime_to_iris_time)
 from improver.nowcasting.utilities import ApplyOrographicEnhancement
+from improver.utilities.stdout_trap import Capture_StdOut
+
+# PySteps is so proud of itself - trap the routine StdOut message on import
```
Comment (arh89) [inline lib/improver/nowcasting/pysteps_advection.py]: Printing when importing!?!

File: lib/improver/utilities/stdout_trap.py (hunk 2 above)
Comment (arh89) [reply lib/improver/utilities/stdout_trap.py]: (In slower time, let's consider putting a PR up to pysteps to address this)

File: lib/improver/utilities/stdout_trap.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+from io import StringIO
+import sys
+
+
+class Capture_StdOut(list):
```
Comment (benfitzpatrick) [inline lib/improver/utilities/stdout_trap.py]: Naming pattern?

File: lib/improver/nowcasting/pysteps_advection.py (hunk 5 above)
Comment (arh89) [reply lib/improver/nowcasting/pysteps_advection.py]: Actually I've just had a look - could someone point out where this print is coming from? (I can't see it at a glance)

Comment (cgsandford) [reply lib/improver/nowcasting/pysteps_advection.py]: Ah I missed this.  PySteps is of course rightly proud - but we should probably rephrase this comment!!!

Comment (cgsandford) [reply lib/improver/nowcasting/pysteps_advection.py]: And yeah.  It says "Look I found a pystepsrc file at </user/specific/path/>!"  Or words to that effect.  Which is probably handy for command-line interactivity, but not so much for real-time running.

File: lib/improver/utilities/stdout_trap.py (hunk 2 above)
Comment (arh89) [reply lib/improver/utilities/stdout_trap.py]: (Actually, can't we resolve this by just having a pysteps config file with `silent_import` as True?)

Comment (MoseleyS) [reply lib/improver/utilities/stdout_trap.py]: I can't find `silent_import` as an option to PySteps, in fact I can't find the string `silent` anywhere in the PySteps code base. Where did you get this suggestion from?

File: lib/improver/nowcasting/pysteps_advection.py (hunk 1 above)
Comment (MoseleyS) [reply lib/improver/nowcasting/pysteps_advection.py]: I take your point about PySteps being designed for precipitation rate, but I can think of fields I would like to advect to produce a nowcast that are not precipitation but are closely linked, e.g. cloud, wind gust. I think it is important to use the same advection method for these so I expect that option to be available. My understanding is that the units are only important to the Optical Flow algorithm, the advection shouldn't care.

File: lib/improver/utilities/stdout_trap.py (hunk 2 above)
Comment (MoseleyS) [reply lib/improver/utilities/stdout_trap.py]: Aaron has showed me that this option is in version 1.1.0 of PySteps, while we are using vesion 1.0.0. I'm not sure when we want to upgrade our installed version.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/improver/utilities/stdout_trap.py  1
- lib/improver/cli/nowcast_extrapolate.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667&bid=15433284)

File: lib/improver/nowcasting/pysteps_advection.py [hunk 7]
```diff
@@ -197,12 +210,18 @@ def _generate_forecast_cubes(self, all_forecasts):
...
+                    cube, self.orogenh)
+
+            # Update meta-data
+            amend_attributes(cube, attributes_dict)
+            set_history_attribute(cube, "Nowcast")
             forecast_cubes.append(cube)
         return forecast_cubes
 
-    def process(self, initial_cube, ucube, vcube, orographic_enhancement):
+    def process(self, initial_cube, ucube, vcube, orographic_enhancement,
```
Comment (benfitzpatrick) [inline lib/improver/nowcasting/pysteps_advection.py]: Codacy Issue found: [Dangerous default value {} as argument](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667)

File: lib/improver/nowcasting/pysteps_advection.py (hunk 4 above)
Comment (MoseleyS) [reply lib/improver/nowcasting/pysteps_advection.py]: Moved.

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/stdout_trap.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667&bid=15433284)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/stdout_trap.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667&bid=15433284)

File: lib/improver/nowcasting/pysteps_advection.py (hunk 1 above)
Comment (MoseleyS) [reply lib/improver/nowcasting/pysteps_advection.py]: I couldn't find anything in the PySteps extrapolate method that said what the units of the input array should be, but there isn't for any of the other fields either. As the input field is named "precip", I have removed this option for now.

File: tests/improver-nowcast-extrapolate/04-basic_no_orographic_enhancement.bats (hunk 3 above)
Comment (MoseleyS) [reply tests/improver-nowcast-extrapolate/04-basic_no_orographic_enhancement.bats]: It's gone.

File: lib/improver/nowcasting/pysteps_advection.py (hunk 5 above)
Comment (MoseleyS) [reply lib/improver/nowcasting/pysteps_advection.py]: I've changed the comment.

File: tests/improver-nowcast-extrapolate/04-basic_no_orographic_enhancement.bats (hunk 3 above)
Comment (MoseleyS) [reply tests/improver-nowcast-extrapolate/04-basic_no_orographic_enhancement.bats]: Removed.

File: lib/improver/nowcasting/pysteps_advection.py [hunk 8]
```diff
@@ -189,21 +178,27 @@ def _set_up_output_cubes(self, all_forecasts):
...
+                of the output cube.
 
         Returns:
             forecast_cubes (list):
                 List of iris.cube.Cube instances containing forecasts at all
                 required lead times, and conforming to the IMPROVER metadata
                 standard.
         """
+        if not attributes_dict:
+            attributes_dict = {}
```
Comment (cgsandford) [inline lib/improver/nowcasting/pysteps_advection.py]: Should this be `if attributes_dict is None`?

Comment (arh89) [reply lib/improver/nowcasting/pysteps_advection.py]: (An empty dictionary is treated as false-y, so unless you explicitly want to distinguish between `None` and an empty dictionary, `if not x` is fine.)

Comment (MoseleyS) [reply lib/improver/nowcasting/pysteps_advection.py]: My thinking was that a user could pass in anything that evaluates as False, so any of the following are ok:
- `None`
- `{}`
- `[]`
- `False`

Comment (cgsandford) [reply lib/improver/nowcasting/pysteps_advection.py]: OK.

**Comment by Katie-Howard** [comment]
I have managed to run the tests (I had made a slight mistake in my site-init) and now they all run fine so I'm happy with this now. 👍 

As for the acceptance criteria:

- [x] Create a nowcasting feature branch 

- [x] In this branch: Write a plugin that wraps the pysteps functionality Actually call the plugin (aka "Implement pysteps version of semi-Lagrangian backwards advection within "CreateExtrapolationForecast" plugin").  Could replace call to "AdvectField" or create a separate wrapper plugin and CLI. 

- [x] Demonstrate this works on IMPROVER-style data / NetCDF files (see IMPRO-1308 for details

File: lib/improver/utilities/stdout_trap.py (hunk 2 above)
Comment (MoseleyS) [reply lib/improver/utilities/stdout_trap.py]: I've replaced the class with a contextmanager function as requested by @arh89

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- lib/improver/utilities/redirect_stdout.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667&bid=15433284)

File: lib/improver/utilities/redirect_stdout.py [hunk 9]
```diff
@@ -28,30 +28,17 @@
 # CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
+"""Contains a class to trap stdout. Deprecated from PySteps v1.1.0 in favour
+of adding "silent_import=True" to the pyconfig file."""
 
-. $IMPROVER_DIR/tests/lib/utils
+from io import StringIO
```
Comment (benfitzpatrick) [inline lib/improver/utilities/redirect_stdout.py]: Codacy Issue found: [Unused StringIO imported from io](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667)

File: lib/improver/utilities/redirect_stdout.py [hunk 10]
```diff
@@ -28,30 +28,17 @@
...
-  [[ "$status" -eq 0 ]]
-
-  improver_check_recreate_kgo "output.nc" $KGO
-
-  # Run nccmp to compare the output and kgo.
-  improver_compare_output "$TEST_DIR/output.nc" \
-      "$IMPROVER_ACC_TEST_DIR/$KGO"
-}
+@contextlib.contextmanager
+def redirect_stdout(target=None):
```
Comment (benfitzpatrick) [inline lib/improver/utilities/redirect_stdout.py]: Codacy Issue found: [Missing function or method docstring](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- lib/improver/utilities/redirect_stdout.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metoppv_tech/improver/pullRequest?prid=4597667&bid=15433284)

**Pull request merged**